libcontainer/container_linux.go

```diff
@@ -518,57 +518,116 @@ func (c *Container) shouldSendMountSources() bool {
 		return false
 	}
 
-	// We need to send sources if there are bind-mounts.
+	// We need to send sources if there are non-idmap bind-mounts.
 	for _, m := range c.config.Mounts {
-		if m.IsBind() {
+		if m.IsBindOnly() {
 			return true
 		}
 	}
 
 	return false
 }
 
-func (c *Container) newInitProcess(p *Process, cmd *exec.Cmd, messageSockPair, logFilePair filePair) (*initProcess, error) {
-	cmd.Env = append(cmd.Env, "_LIBCONTAINER_INITTYPE="+string(initStandard))
-	nsMaps := make(map[configs.NamespaceType]string)
-	for _, ns := range c.config.Namespaces {
-		if ns.Path != "" {
-			nsMaps[ns.Type] = ns.Path
+// shouldSendIdmapSources says whether the child process must setup idmap mounts with
+// the mount_setattr already done in the host user namespace.
+func (c *Container) shouldSendIdmapSources() bool {
+	// For the time being we require userns to be in use.
+	if !c.config.Namespaces.Contains(configs.NEWUSER) {
+		return false
+	}
+
+	// nsexec.c mount_setattr() requires CAP_SYS_ADMIN in:
+	// * the user namespace the filesystem was mounted in
+	// * the user namespace we're trying to idmap the mount to
+	// * the owning user namespace of the mount namespace you're currently located in.
+	//
+	// See the comment from Christian Brauner:
+	//	https://github.com/opencontainers/runc/pull/3717#discussion_r1103607972
+	//
+	// Let's just rule out rootless, we don't have those permission in the
+	// rootless case.
+	if c.config.RootlessEUID {
+		return false
+	}
+
+	// We need to send sources if there are idmap bind-mounts.
+	for _, m := range c.config.Mounts {
+		if m.IsIDMap() {
+			return true
 		}
 	}
-	_, sharePidns := nsMaps[configs.NEWPID]
-	data, err := c.bootstrapData(c.config.Namespaces.CloneFlags(), nsMaps, initStandard)
+
+	return false
+}
+
+func cmdAddEnvFds(cmd *exec.Cmd, env string, fds []int) error {
+	fdsJSON, err := json.Marshal(fds)
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("Error creating %v: %w", env, err)
 	}
+	cmd.Env = append(cmd.Env, env+"="+string(fdsJSON))
+	return nil
+}
 
-	if c.shouldSendMountSources() {
-		// Elements on this slice will be paired with mounts (see StartInitialization() and
-		// prepareRootfs()). This slice MUST have the same size as c.config.Mounts.
-		mountFds := make([]int, len(c.config.Mounts))
-		for i, m := range c.config.Mounts {
-			if !m.IsBind() {
-				// Non bind-mounts do not use an fd.
-				mountFds[i] = -1
-				continue
-			}
+func (c *Container) sendSources(cmd *exec.Cmd, messageSockPair filePair) error {
+	if !c.shouldSendMountSources() && !c.shouldSendIdmapSources() {
+		return nil
+	}
 
+	// Elements on these slices will be paired with mounts (see StartInitialization() and
+	// prepareRootfs()). These slices MUST have the same size as c.config.Mounts.
+	mountFds := make([]int, len(c.config.Mounts))
+	idmapFds := make([]int, len(c.config.Mounts))
+	for i, m := range c.config.Mounts {
+		// The -1 fd is ignored later.
+		mountFds[i] = -1
```

Comment:
Not critical, but my IDE is warning that  `mountFds` here is shadowing the `mountFds` type that was added in libcontainer/init_linux.go. Perhaps best to pick a different name for one of them.

Reply:
I don't see this in my vim, but I think it might be solved as this var doesn't exist anymore. But please confirm :)